[flang] Lower procedure pointer components

jeanPerier opened the pull request:
Lower procedure pointer components, except in the context of structure constructor (left TODO).

Procedure pointer components lowering share most of the lowering logic of procedure poionters with the following particularities:
- They are components, so an hlfir.designate must be generated to retrieve the procedure pointer address from its derived type base.
- They may have a PASS argument. While there is no dispatching as with type bound procedure, special care must be taken to retrieve the derived type component base in this case since semantics placed it in the argument list and not in the evaluate::ProcedureDesignator.

These components also bring a new level of recursive MLIR types since a fir.type may now contain a component with an MLIR function type where one of the argument is the fir.type itself. This required moving the "derived type in construction" stackto the converter so that the object and function type lowering utilities share the same state (currently the function type utilty would end-up creating a new stack when lowering its arguments, leading to infinite loops). The BoxedProcedurePass also needed an update to deal with this recursive aspect.

jeanPerier committed:
[flang] Lower procedure pointer components

Lower procedure pointer components, except in the context of
structure constructor (left TODO).

Procedure pointer components lowering share most of the lowering
logic of procedure poionters with the following particularities:
- They are components, so an hlfir.designate must be generated to
  retrieve the procedure pointer address from its derived type base.
- They may have a PASS argument. While there is no dispatching as with
  type bound procedure, special care must be taken to retrieve the
  derived type component base in this case since semantics placed it
  in the argument list and not in the evaluate::ProcedureDesignator.

These components also bring a new level of recursive MLIR types since a
fir.type may now contain a component with an MLIR function type where
one of the argument is the fir.type itself. This required moving the
"derived type in construction" stackto the converter so that the object
and function type lowering utilities share the same state (currently the
function type utilty would end-up creating a new stack when lowering its
arguments, leading to infinite loops). The BoxedProcedurePass also
needed an update to deal with this recursive aspect.

psteinfeld reviewed (approved):
All builds and tests correctly.

clementval reviewed (approved):
LGTM

DanielCChen commented on flang/lib/Lower/ConvertCall.cpp:
```diff
@@ -165,20 +166,29 @@ fir::ExtendedValue Fortran::lower::genCallOpAndResult(
+      Fortran::lower::SomeExpr expr{*procDesignator};
+      fir::ExtendedValue loweredProc =
+          converter.genExprAddr(loc, expr, stmtCtx);
+      funcPointer = fir::getBase(loweredProc);
```
This is a clarification question: 
For referencing a non-component procedure pointer, how its "value" is loaded? Does `getBase` do that?

jeanPerier replied:
`genExprAddr` does it. For Pointers and Allocatables, the behavior of genExtAddr is to return the base address, not the address of the pointer/descriptor. So for procedure pointers, it will return the function address.

DanielCChen replied:
Thanks for the explanation. LGTM.

jeanPerier committed:
turn an assert into TODO

jeanPerier committed:
Fix character procedure pointer component

The result argument is not placed in the list yet when looking for the
passed argument to lower the procedure reference. This caused an assert
to fire. Add an entry point to get the passed argument without checking
if the other arguments have been lowered yet.

Also visit fir.llvm_ptr type when rewriting fir.boxproc type (needed to
when a derived type with a procedure pointer component is "captured" in
a host procedure because it is used in an internal procedure).

jeanPerier committed:
convert fir.emboxproc result type

jeanPerier committed:
drop type conversion markers when deconstructing names

Otherwise, the deconstruction will complain about unknown markers.

jeanPerier committed:
Merge branch 'main' into jpr-proc-pointer-components
